### ISSUE TITLE
RECAP Archive Has Duplicate, Broken Dockets for Some Federal Appellate Cases

### ISSUE THREAD

**Issue by stevenxdavis** [freelawproject/courtlistener#1376]
TL;DR, there are approximately 1,500 duplicate docket pages in the RECAP Archive where the "View on PACER" hyperlink does nothing but still costs the user 10 cents.

The RECAP Archive on Court Listener has duplicate dockets for some federal appellate cases, and one of them will contain a broken hyperlink.

This isn't a database problem per se - the database is clearly built to allow multiple dockets with the same docket number in the same jurisdiction. For instance, there are 6 database entries for docket number 17-72917 in the 9th Circuit:

https://www.courtlistener.com/api/rest/v3/dockets/?court__id=ca9&docket_number=17-72917

Note that only one of them has a value for pacer_case_id and the rest have empty strings. In the RECAP Archive, this one "full" docket lists the filings and shows up in a search:

https://www.courtlistener.com/docket/6179936/in-re-usa-v-usdc-casf/
https://www.courtlistener.com/?type=r&q=&type=r&order_by=score%20desc&docket_number=17-72917&court=ca9

These remaining URLs lead to dummy pages:

https://www.courtlistener.com/docket/6218978/united-states-v-usdc-casf/
https://www.courtlistener.com/docket/6225331/in-re-united-states/
https://www.courtlistener.com/docket/6226308/in-re-united-states/
https://www.courtlistener.com/docket/6228430/in-re-united-states/
https://www.courtlistener.com/docket/6246250/in-re-united-states/

So far, so good. 

But there are some cases (I found approximately 1,500) with two full dockets in the RECAP Archive that both show up in a search:

https://www.courtlistener.com/docket/19196/jamal-adams-v-postmates-inc/
https://www.courtlistener.com/docket/20414/jamal-adams-v-postmates-inc/
https://www.courtlistener.com/?type=r&q=&type=r&order_by=score%20desc&docket_number=19-17362&court=ca9

As far as I can tell, the problem arises when a case has two dockets in the database with the same docket number but different PACER IDs:

https://www.courtlistener.com/api/rest/v3/dockets/?court__id=ca9&docket_number=19-17362

Note that there can be even more docket entries that have empty pacer_case_id values and don't show up in the RECAP Archive:

https://www.courtlistener.com/api/rest/v3/dockets/?court__id=ca9&docket_number=19-15072 (three entries)
https://www.courtlistener.com/?type=r&q=&type=r&order_by=score%20desc&docket_number=19-15072&court=ca9 (two search results)
https://www.courtlistener.com/docket/16363807/state-of-california-v-the-little-sisters-of-the-poor/ (one dummy docket)

What brought this to my attention is that on the "mismatched" docket, the "View on PACER" link will be broken. For instance, on this page:

https://www.courtlistener.com/docket/8341441/commonwealth-of-pennsylvania-v-president-united-states/

"View on PACER" links to:

https://ecf.ca3.uscourts.gov/n/beam/servlet/TransportRoom?servlet=CaseSummary.jsp&caseNum=105152&incOrigDkt=Y&incDktEntries=Y

When this page loads, you'll get "105152 Case Number Not Found" and you'll be out 10 cents, which is a real bummer and a drag. This is because PACER's CaseSummary.jsp requires caseNum (i.e. docket_number) to be in YY-NNNNN format.

Note on the same page that you can click on any document's "Buy on PACER" link and get the actual document, e.g.:

https://ecf.ca3.uscourts.gov/docs1/003012864689?caseid=105152

This is because PACER's ShowDoc servlet can handle PACER's internal caseid values (pacer_case_id) or the court's docket numbers (docket_number).

When I was going through this over the summer, I downloaded the database to try and figure this out, and I attached a spreadsheet of all the cases I found with this problem (as of June 21). The spreadsheet has all database entries from federal appellate courts where the docket_number and pacer_case_id have different values and pacer_case_id is nonempty; for each entry in this spreadsheet, there will also be a second entry in the database where the docket_number matches the pacer_case_id. These cases tend to have fairly high profiles, so I suspect the database is being updated by two (or more) different processes that aren't merging docket changes correctly.

The simplest fix for the hyperlinking problem would be to have "View on PACER" link use the docket_number and not the pacer_case_id, ensuring that the format will match what CaseSummary.jsp expects. If nothing else, users won't inadvertently lose 10 cents by clicking the links.

It will be a lot more complicated to remove the duplicates and ensure more aren't created; I think you'll have to figure out where they are coming from.

I'm a second-year law student and my last computer science class was in 2009, so I hope this all makes some sense. Let me know if anything is unclear and I'll see if I can explain further.

[COA_Duplicates.xlsx](https://github.com/freelawproject/courtlistener/files/5079832/COA_Duplicates.xlsx)

**Comment by mlissner** [comment]
I suspect this might be fixed or at least no longer getting worse. @albertisfu, can you take a look at this when you get a minute? Maybe we can close it out.

**Comment by albertisfu** [comment]
I have gone through the examples described in this issue, and these are my findings:

**About duplicated dockets like the ones here:**
https://www.courtlistener.com/api/rest/v3/dockets/?court__id=ca9&docket_number=17-72917

Most of them were created based on an opinion, for example: [https://www.courtlistener.com/docket/6228430/in-re-united-states/](https://www.courtlistener.com/docket/6228430/in-re-united-states/)

This duplication issue should now be resolved after we applied a lookup to find an existing docket when adding opinion #2596 from `cl_scrape_oral_arguments`, `cl_scrape_opinions`, and `harvard_opinions`.

In addition to the fix applied in #2511, we will now generate `docket_number_core` only for known formats.

Just this one seems to be different:
https://www.courtlistener.com/docket/6218978/united-states-v-usdc-casf/
It has the same docket number 17-72917, but its case name is different from the others. This was added for an Oral Argument. Is it the same as the previous case (in-re-united-states)? If it's not the same case, cases like this might be a problem since it doesn't have a pacer_case_id. So even with the last fixes, it should have been merged into the wrong docket.

Since all of these dockets don't have a `pacer_case_id` its PACER link uses the `docket_number` and it points to: 
[USA, et al v. USDC-CASF](https://ecf.ca9.uscourts.gov/n/beam/servlet/TransportRoom?servlet=CaseSummary.jsp&caseNum=17-72917&incOrigDkt=Y&incDktEntries=Y) 


**On the following reported dockets:** 
https://www.courtlistener.com/docket/19196/jamal-adams-v-postmates-inc/ 
https://www.courtlistener.com/docket/20414/jamal-adams-v-postmates-inc/

Both of them are linked to an opinion, so the docket was created with the opinion.

e.g: https://www.courtlistener.com/opinion/2643541/daniel-saunders-v-king-county/
The opinion and the docket were created on Oct. 29, 2014.

However the docket now has a different case name, so in this case, the problem seems to be related to the scrambled dockets issue on the extension that was fixed by Eduardo last year.

**Regarding broken PACER links on mismatched dockets**, the issue is also fixed after Eduardo fixed appellate PACER links.
e.g: https://www.courtlistener.com/docket/8341441/commonwealth-of-pennsylvania-v-president-united-states/

The problem was that before the PACER link, used always the `caseNum` but used the `pacer_case_id` as value:

https://ecf.ca3.uscourts.gov/n/beam/servlet/TransportRoom?servlet=CaseSummary.jsp&caseNum=105152&incOrigDkt=Y&incDktEntries=Y

Now is solved and it uses the `caseId` when the `pacer_case_id` is available.
https://ecf.ca3.uscourts.gov/n/beam/servlet/TransportRoom?servlet=CaseSummary.jsp&caseId=105152&incOrigDkt=Y&incDktEntries=Y

And the `caseNum` and the `docket_number` when the `pacer_case_id` is not available:
https://ecf.ca9.uscourts.gov/n/beam/servlet/TransportRoom?servlet=CaseSummary.jsp&caseNum=17-72917&incOrigDkt=Y&incDktEntries=Y

The cleanup of appellate dockets that had the `docket_number` into the `pacer_case_id` field also helped to solve this issue.

**Also checked the `xls` attached that mentioned other dockets with broken links, the problem is also solved on most of them.**

Just a few exceptions where the docket is still broken, not due to the issues mentioned before, but because the docket has a bad `docket_number `or bad `pacer_case_id`, possibly from its original source.

https://www.courtlistener.com/docket/7296372/united-states-v-levin/
`pacer_case_id`: 16

https://www.courtlistener.com/docket/25105/bushay-v-mcdonnell/
`docket_number`: 05-9

https://www.courtlistener.com/docket/4490126/painters-and-allied-trades-v-forest-pharmaceuticals-inc/
`pacer_case_id`: None

Here we could still clean up **pacer_case_id** values in appellate where the `pacer_case_id` is too short or "None."

So, in brief, the errors reported in this issue have been covered by recent fixes and should no longer continue to happen. The duplicated dockets, which were mostly created based on an opinion, are still duplicated in the system, so they might also need to be cleaned up.

**Comment by mlissner** [comment]
Great:

 - I updated the 12 cases that had "None" as their `pacer_case_id`. 
 - I looked at some of the cases that had short `pacer_case_id` values, and they're legitimate, so I'm not messing with that.

I think that's a wrap here. Anything else I can/should do?

**Comment by albertisfu** [comment]
Perfect, now the PACER link works on cases where they had `None`.
I think the work here is done since the bugs causing the problems reported here have been solved.
